### PR TITLE
[FEATURE] Rule-Based Profiler Configuration and Runtime Arguments Reconciliation Logic

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Refactoring of Rule-Based Profiler codebase for improved importing capability and increased encapsulation.
* Implementation of reconciliation between runtime arguments supplied to running the profiler and original configuration.
* In-line (docstrings) documentation describing the reconciliation logic.
* Unit tests covering a wide variety of argument-configuration reconciliation scenarios.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-489
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/data_context/data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context.py:122]: Are we sure about this? I believe we consciously decided to use `types` since it isn't informative.

I understand the value in consistency but this was a specific design decision made by @donaldheppner. Please confer with him just to get his views here.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule/rule.py]: I appreciate verbosity but this is really hard to read in my opinion.

Also, I don't think we need `list` here or above (you can use the iterator as is)

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule/rule.py]: I'm a firm believer that variable names should reflect the contents but not their type. Naming this dict and then annotating it as Dict feels redundant.

Can we think of a better name for the dictionary representation of the object? Perhaps `serialized_parameter_builders`?

Comment (cdkini) [reply great_expectations/rule_based_profiler/rule/rule.py]: This same note extends to method names (I feel less strongly about those though)

File: great_expectations/rule_based_profiler/types/serializable_builder.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/serializable_builder.py]: This logic feels super susceptible to breakage in my opinion - why not offset the burden of implementation to the child classes and specify the logic there?

I love generalizing and reusing code but I'm wary of catch all code here.

Let's also split up this logic into different lines to help with debugging (`deep_filer_properties_iterable`, `super().to_dict()`, `items()` iteration, all on separate lines)

File: great_expectations/rule_based_profiler/util.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/util.py]: Please add a docstring here and in all other public functions.

Using the `Raises:` section of the Google style guide please!

Comment (cdkini) [reply great_expectations/rule_based_profiler/util.py]: Also, can we simplify this check? We don't need to do any bitwise operators or set conversion in my opinion.

```python
if not isinstance(builder_config, dict) or any(attr in builder_config for attr in ("class_name", "module_name"):
    raise
```

File: tests/conftest.py
Comment (cdkini) [inline tests/conftest.py:5034]: Please add a docstring to clarify what you mean by placeholder_args (and how this differs from a standard use-case)

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py:8]: If you annotate one arg, please ensure you annotate them all 😄

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py:68]: Can this go in a fixture? This single test feels very verbose and we can possibly reuse the data herein.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py]: Should we be using `to_json_dict` here? If that utility is broken, we're relying on that behavior to test another part of the class.

Also, I'd love it if we split this up into `actual` and `expected` and asserted `actual == expected`. It may take up a few more lines but I think it really helps with readability

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py]: Same note here.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py]: And here.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py]: Same note for the remainder of the file

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:629]: Same note about variable naming - I think it's not the best practice to name the variable based on the type.

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:635]: Rather than use dunders here, can we expose this functionality through methods on the builder objects?

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:645]: This is that roundtrip we were discussing - could we please encapsulate it in a Schema level method and provide some more context as to why we're doing this instead of `validate`?

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Instead of an `Optional` list, can we just return an empty list?

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:560]: Let's move this functionality to methods on the parameter builder class

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Let's figure out whether or not `validate` is appropriate here and if not, move the logic to a schema level method as noted in our Slack convos.

File: great_expectations/rule_based_profiler/util.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/util.py]: Per discussion with @donaldheppner in my own PR, I agree with his approach of making methods like return boolean values. Validation should be true/false and an exception should be raised at the same level as function invocation.

**Review by cdkini — CHANGES_REQUESTED** [review]
Left general comments - please let me know if you'd like to discuss.

File: great_expectations/data_context/data_context.py
Comment (alexsherstinsky) [reply great_expectations/data_context/data_context.py:122]: @cdkini We are indeed sure about this for now.  I have an agreement with @donaldheppner to introduce typed objects to `Checkpoint` and `RuleBasedProfiler` instantiation ASAP (as in -- in the soonest follow-on PRs, provided the RBP work is unblocked).  Hence, I am planning to get to it no later than the start of the next sprint (likely sooner).  Thanks.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule/rule.py]: @cdkini Indeed, `list` is not needed (deleted in the next commit).  I would like to keep this verbose style in for now.  Thanks.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule/rule.py]: @cdkini I cleaned up the naming to not have this redundancy (in next commit).  Thanks!

File: great_expectations/rule_based_profiler/types/serializable_builder.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/serializable_builder.py]: @cdkini Let us leave this one for discussion to agree on what makes sense to do here.  There are a few trade-offs.

File: great_expectations/rule_based_profiler/util.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/util.py]: @cdkini I will add the docstring (in the next commit).  Please note that this is not a boolean check, but must raise an exception (as in -- "cannot continue" -- I will document this.  Also, this ins not a bitwise operation (it is the recommended `set` shorthand for `intersection`).

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/util.py]: @cdkini I also moved it to a more appropriate module (in the next commit).

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:635]: @cdkini Let us add this one to our discussion as well.  Thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:645]: @cdkini Let us add this one to our discussion as well.  Thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @cdkini Good idea -- done (in next commit) -- thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:560]: @cdkini Let us add this one to our discussion as well. Thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @cdkini Let us add this one to our discussion as well. Thanks!

File: tests/conftest.py
Comment (alexsherstinsky) [reply tests/conftest.py:5034]: @cdkini I added a docstring.  Thanks.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py:8]: @cdkini I deleted the "lone" type hint (informally, I got used to not including type hints when including fixtures; happy to change it globally should we all move in that direction).  Thanks.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py:68]: @cdkini This is a one-time example (not sure if we need this for any other purposes).  Thanks.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: @cdkini I will fix this once we get past the `Python-3.6 related failures` and before merging.  Thanks!

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: @cdkini I will fix this once we get past the `Python-3.6 related failures` and before merging.  Thanks!

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: @cdkini I will fix this once we get past the `Python-3.6 related failures` and before merging.  Thanks!

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: @cdkini I will fix this once we get past the `Python-3.6 related failures` and before merging.  Thanks!

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py:629]: @cdkini I fixed this (committed).

File: great_expectations/data_context/data_context.py
Comment (donaldheppner) [reply great_expectations/data_context/data_context.py:122]: but if we had it in `.config` before, why are we moving it to `.types.base` now?

File: great_expectations/rule_based_profiler/config/__init__.py
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/config/__init__.py:1]: this was good! Why delete?

File: great_expectations/rule_based_profiler/domain_builder/domain_builder.py
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/domain_builder/domain_builder.py]: Why is this serializable?

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/rule_based_profiler/config/__init__.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/config/__init__.py:1]: @donaldheppner It made more sense to keep this and other classes under `type`, hence I moved this to `base` just like it is in the `data_context` scheme.  And right now, all imports work, and no circular dependencies.

File: great_expectations/rule_based_profiler/domain_builder/domain_builder.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/domain_builder.py]: @donaldheppner Because it implements `to_json_dict()`.

File: great_expectations/data_context/data_context.py
Comment (alexsherstinsky) [reply great_expectations/data_context/data_context.py:122]: @donaldheppner Please see my other response -- for now, this makes all model-type models have a "good home" in "types", hence, "config" was moved to `base.py` just like it is under the `data_context` directory organization.

File: tests/rule_based_profiler/test_rule_based_profiler.py
Comment (alexsherstinsky) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: @cdkini All done.  (Also updated the Marshmallow Schema roundtrip comment for better clarity.)  Thanks.

File: great_expectations/checkpoint/configurator.py
Comment (cdkini) [inline great_expectations/checkpoint/configurator.py:162]: I appreciate this comment 🙇🏽

File: great_expectations/data_context/data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context.py:122]: Will defer to @donaldheppner here.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule/rule.py:166]: Why `if parameter_builders`? Is it not guaranteed by the `get` method?

Thoughts on making this return an empty list instead of `None`? I'd love to move away from nulls whenever we can.

Comment (cdkini) [reply great_expectations/rule_based_profiler/rule/rule.py:166]: I see that the `get` can return an empty dict. Let's just do this instead?:

```python
return list(parameter_builders.values())
```

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Let's separate this out into different lines incase we get key errors?

Also, is this another instance where we can get rid of `_dict` suffix?

File: great_expectations/rule_based_profiler/config/__init__.py
Comment (donaldheppner) [reply great_expectations/rule_based_profiler/config/__init__.py:1]: "type" is not a meaningful qualifier -- everything is a Type! I'd really like to preserve the existing, more descriptive, semantic module name we had before, or understand the reason we need to revert.

File: great_expectations/rule_based_profiler/types/builder.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/builder.py:11]: This is to be removed right? Can we make a note about that?

File: tests/conftest.py
Comment (cdkini) [inline tests/conftest.py]: This should be 3.7 no?

File: tests/conftest.py
Comment (cdkini) [inline tests/conftest.py]: Can you add a note regarding what's breaking? (dataclasses)

File: tests/conftest.py
Comment (cdkini) [inline tests/conftest.py:5039]: I think we may want to rethink this pattern but it is good for now!

**Review by cdkini — COMMENTED** [review]
Left a few general comments I'd love to see resolved before we merge! Nothing crazy mainly documentation/comments.

File: tests/conftest.py
Comment (alexsherstinsky) [reply tests/conftest.py]: @cdkini It is *typo* -- fixed by now.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule/rule.py:166]: @cdkini This seems a bit of an overkill -- it is `Pythonic` do test the truthiness of a collection.  So I would want to leave it as is.

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @cdkini I will change the name (in the next commit).  No need to guard against they key errors -- guaranteed to be non-NULL.

File: great_expectations/rule_based_profiler/types/builder.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/builder.py:11]: @cdkini Done.

File: tests/conftest.py
Comment (alexsherstinsky) [reply tests/conftest.py]: @cdkini Done.

File: great_expectations/rule_based_profiler/rule/rule.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/rule/rule.py:166]: Due to being nullable in the schema, we deem this acceptable!

**Review by cdkini — APPROVED** [review]
LGTM!

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**
